### PR TITLE
Add static DHCPv4 leases settings

### PR DESCRIPTION
**Pull request by DL6ER**
**By submitting this pull request, I confirm the following (please check boxes, eg [X] - no spaces) _Failure to fill the template will close your PR_:**

***Please submit all pull requests against the `development` branch. Failure to do so will delay or deny your request***

- [X] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md).
- [X] I have checked that [another pull request](https://github.com/pi-hole/pi-hole/pulls) for this purpose does not exist.
- [X] I have considered, and confirmed that this submission will be valuable to others.
- [X] I accept that this submission may not be used, and the pull request closed at the will of the maintainer.
- [X] I give this submission freely, and claim no ownership to its content.

**How familiar are you with the codebase?:** 10

---

Add static DHCPv4 leases settings to Settings page. Batch-processing is not supported (add one entry at a time).

Backend changes: https://github.com/pi-hole/pi-hole/pull/1170

screenshot at 2017-01-25 11-59-39

_This template was created based on the work of [`udemy-dl`](https://github.com/nishad/udemy-dl/blob/master/LICENSE)._

### CODE REVIEW

**Comment by darkrain-nl** [comment]
Can you please explain why the MAC address is mandatory?

DNSMasq also works nice if you specify a hostname and the IP it should get, very handy if you don't know the MAC but do know the hostname...

**Comment by DL6ER** [comment]
From `dnsmasq` `man`-page
> dhcp-host=lap,192.168.0.199 tells dnsmasq to always allocate the machine lap the IP address 192.168.0.199.

We cannot detect if multiple hosts are asking for the same address (because they have the same host name, like `debian`) and that could lead to issues.

**Comment by darkrain-nl** [comment]
That's true, but that's up to the administrator to make sure hostnames are unique.
In fact this is the first thing I do when setting up a new device...

**Comment by DL6ER** [comment]
Well, still people might bring laptops and/or smartphones to your WiFi and you might give them the password without thinking that their devices could have a bad host name.

Imagine you have a Win phone whose host name is just `winphone` and you cannot change it (not sure if they do it). Any second `winphone` coming to your network would cause trouble, as `dnsmasq` will do *nothing* to prevent IP collisions from happening.

**Comment by darkrain-nl** [comment]
I agree, in that case leave it like it is.
I only have one or two devices configured with only hostname, and that's still possible by editing the files manually...

**Comment by DL6ER** [comment]
> that's still possible by editing the files manually...

I think we should really leave that for professionals. However, I will have to adapt the translation routines so your results will still be display correctly.

**Comment by darkrain-nl** [comment]
> I think we should really leave that for professionals.

I am ;-)

> However, I will have to adapt the translation routines so your results will still be display correctly.

That would be nice!

**Comment by DL6ER** [comment]
screenshot at 2017-01-25 13-07-57

**Comment by darkrain-nl** [comment]
Looks good!

File: scripts/pi-hole/php/savesettings.php [hunk 1]
```diff
@@ -316,6 +358,76 @@ function validDomain($domain_name)
...
+					if(!validMAC($mac))
+					{
+						$error .= "MAC address (".htmlentities($mac).") is invalid!<br>";
+					}
+					$mac = strtoupper($mac);
+
+					if(!strlen($error))
+					{
+						exec("sudo pihole -a removestaticdhcp ".$mac);
+						$success .= "The static address with MAC address ".htmlentities($mac)." has been added";
```
Comment (Mcat12) [inline scripts/pi-hole/php/savesettings.php]: added -> removed

**Comment by Mcat12** [comment]
Approved. DON'T MERGE UNTIL POINT RELEASE IS OUT.

**Pull request merged**